datafusion/core/src/datasource/listing/path.rs

```diff
@@ -0,0 +1,217 @@
+    /// an absolute path on the local filesystem, returning an error if it does not exist,
+    /// and converted to a [file URI]
+    ///
+    /// The remaining string will be interpreted as a [`glob::Pattern`] and used as a
+    /// filter when listing files from object storage
+    ///
+    /// [file URI]: https://en.wikipedia.org/wiki/File_URI_scheme
+    pub fn parse(s: impl AsRef<str>) -> Result<Self> {
+        let s = s.as_ref();
+        Ok(match Url::parse(s) {
+            Ok(url) => Self { url, glob: None },
+            Err(url::ParseError::RelativeUrlWithoutBase) => {
+                let (prefix, glob) = match split_glob_expression(s) {
+                    Some((prefix, glob)) => {
+                        let glob = Pattern::new(glob)
+                            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+                        (prefix, Some(glob))
+                    }
+                    None => (s, None),
+                };
+
+                let path = std::path::Path::new(prefix).canonicalize()?;
+                let url = match path.is_file() {
+                    true => Url::from_file_path(path).unwrap(),
+                    false => Url::from_directory_path(path).unwrap(),
+                };
+
+                Self { url, glob }
+            }
+            Err(e) => return Err(DataFusionError::External(Box::new(e))),
+        })
+    }
+
+    /// Returns the URL scheme
+    pub fn scheme(&self) -> &str {
+        self.url.scheme()
+    }
+
+    /// Returns the path as expected by [`ObjectStore`]
+    ///
+    /// In particular for file scheme URLs, this has a leading `/`
+    /// and describes an absolute path on the local filesystem
+    ///
+    /// For other URLs, this also contains the host component
+    /// and lacks a leading `/`
+    ///
+    /// TODO: Handle paths consistently (#2489)
+    fn prefix(&self) -> &str {
+        match self.scheme() {
+            "file" => self.url.path(),
+            _ => &self.url[url::Position::BeforeHost..url::Position::AfterPath],
+        }
+    }
+
+    /// Strips the prefix of this [`ListingTableUrl`] from the provided path, returning
+    /// an iterator of the remaining path segments
+    pub(crate) fn strip_prefix<'a, 'b: 'a>(
+        &'a self,
+        path: &'b str,
+    ) -> Option<impl Iterator<Item = &'b str> + 'a> {
+        // Ignore empty path segments
+        let diff = itertools::diff_with(
+            path.split('/').filter(|s| !s.is_empty()),
+            self.prefix().split('/').filter(|s| !s.is_empty()),
+            |a, b| a == b,
+        );
+
+        match diff {
+            // Match with remaining
+            Some(itertools::Diff::Shorter(_, subpath)) => Some(subpath),
+            _ => None,
+        }
+    }
+
+    /// List all files identified by this [`ListingTableUrl`] for the provided `file_extension`
+    pub(crate) fn list_all_files<'a>(
```

Comment:
This combines the various different ObjectStore listing methods into a single method. This is clearer and more consistent